### PR TITLE
GH-1411 AmChart Fix

### PR DESCRIPTION
**Pull request by mlzummo**
Fixes SerialObj Handling and Persisting

Fixes GH-1411

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

**Comment by mlzummo** [comment]
not finished yet

File: src/amchart/CommonSerial.js [hunk 1]
```diff
@@ -172,48 +170,30 @@
...
+            yAxis._owningWidget = this;
+            yAxis._context = this;
+            this.yAxes()[idx] = yAxis;
+        }
+        return this.yAxes()[idx]
     };
 
     CommonSerial.prototype.formatData = function (d) {
-        switch (this.xAxis()[0].axisType()) {
+        switch (this.xAxes()[0].axisType()) {
```
Comment (mlzummo) [inline src/amchart/CommonSerial.js]: for true multi axis obj support wont i need to loop this?

**Comment by mlzummo** [comment]
@jbrundage can you please to check and see if this works in your env?

**Comment by mlzummo** [comment]
@GordonSmith not sure why these errors are occurdig ... eveyrything looks fine in dermatology

**Comment by mlzummo** [comment]
its ready i think

File: src/amchart/CommonSerial.js [hunk 2]
```diff
@@ -26,27 +26,21 @@
         this._colorObj = {};
         this._selectionObj = {};
 
-        if (!this.xAxis().length) {
-            this.xAxis(0);
-            this._xAxis = this.xAxis()[0];
+        this._xAxis = new Axis();
+        this._xAxis._owningWidget  = this;
+        this._xAxis._context  = this;
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonXY.js [hunk 3]
```diff
@@ -19,21 +19,22 @@
...
+        this._yAxis._owningWidget  = this;
+        this._yAxis._context  = this;
     }
+
     CommonXY.prototype = Object.create(HTMLWidget.prototype);
     CommonXY.prototype.constructor = CommonXY;
     CommonXY.prototype._class += " amchart_CommonXY";
 
+    CommonXY.prototype.Axis = Axis;
+
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: Remove

File: test/amchartFactory.js [hunk 4]
```diff
@@ -82,10 +83,8 @@
                     ;
                     col.xAxis(0).axisType("ordinal");
                     col.yAxis(0).axisType("linear");
-                    col.yAxis(1)
-                        .axisType("linear")
-                        .position("right")
-                    ;
+                    col.yAxis(1).axisType("linear");
+                    col.yAxis(1).position("right")  ;
```
Comment (GordonSmith) [inline test/amchartFactory.js]: Why this change?  Not needed.

**Comment by mlzummo** [comment]
@GordonSmith made suggested changes

File: src/amchart/CommonSerial.js [hunk 5]
```diff
@@ -172,48 +162,29 @@
...
+        }
+        return this.xAxes()[idx];
     };
 
     CommonSerial.prototype.yAxis = function (idx) {
-        return this.Axes("y", idx);
+        if (!this.yAxes()[idx]) {
+            var yAxis = new Axis();
+            yAxis._owningWidget = this;
+            yAxis._context = this;
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonSerial.js [hunk 6]
```diff
@@ -583,6 +555,19 @@
...
+        if (this.xAxes().length === 0) {
+            this.xAxes().push(this._xAxis);
+        }
+        if (this.yAxes().length === 0) {
+            this.yAxes().push(this._yAxis);
+        }
+        if (this.y2().length && this.yAxes().length === 1) {
+            var y2Axis = new Axis();
+            y2Axis.owningWidget  = this;
+            y2Axis._context  = this;
```
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonXY.js [hunk 7]
```diff
@@ -19,21 +19,22 @@
         this._prevDataUpdated = -1;
         this._columnsUpdated = 0;
         this._prevColumnsUpdated = -1;
+        
+        this._xAxis = new Axis();
+        this._xAxis._owningWidget  = this;
+        this._xAxis._context  = this;
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonXY.js [hunk 8]
```diff
@@ -134,43 +134,24 @@
...
-        axe.call(this, currentAxes);
-        return axis;
-    };
-
     CommonXY.prototype.xAxis = function (idx) {
-        return this.Axes("x", idx);
+        if (!this.xAxes()[idx]) {
+            var xAxis = new Axis();
+            xAxis._owningWidget = this;
+            xAxis._context = this;
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonXY.js [hunk 9]
```diff
@@ -134,43 +134,24 @@
...
+        }
+        return this.xAxes()[idx];
     };
 
     CommonXY.prototype.yAxis = function (idx) {
-        return this.Axes("y", idx);
+        if (!this.yAxes()[idx]) {
+            var yAxis = new Axis();
+            yAxis._owningWidget = this;
+            yAxis._context = this;
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: remove all instances of _context as it was replaced with _owningWidget

File: src/amchart/CommonXY.js [hunk 10]
```diff
@@ -19,21 +19,22 @@
...
-            this._xAxis = this.xAxis()[0];
-
-        }
-        if (!this.yAxis().length) {
-            this.yAxis(0);
-            this._yAxis = this.yAxis()[0];
-        }
+        this._yAxis = new Axis();
+        this._yAxis._owningWidget  = this;
+        this._yAxis._context  = this;
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: remove all instances of _context as it was replaced with _owningWidget

**Comment by GordonSmith** [comment]
@mlzummo You missed a bunch of comments.

File: src/amchart/CommonSerial.js (hunk 1 above)
Comment (GordonSmith) [inline src/amchart/CommonSerial.js]: @mlzummo  where is your test case?

File: test/amchartFactory.js (hunk 4 above)
Comment (mlzummo) [inline test/amchartFactory.js]: u asked me to put it back ... so i did ... it will work either way lets just leave it ... it shows u can do it both ways
